scripts/download.pl

```diff
@@ -284,7 +284,7 @@ sub trash
 		cleanup();
 		exit 0 if $sum eq $file_hash;
 
-		die "Hash of the local file $filename does not match (file: $sum, requested: $file_hash) - deleting download.\n";
+		warn "Hash of the local file $filename does not match (file: $sum, requested: $file_hash) - deleting download.\n";
```

Comment:
NACK

Reply:
tell me your thoughts on it

AFAIK, it just tries the other mirrors too and then dies the same way

if I were to touch the filename to be downloaded and then run the download it will only exit without attempting a download

Reply:
by the way, the script dies saying "deleting download" before the line where it deletes the file, so it doesn't even do what it says it does

Reply:
> by the way, the script dies saying "deleting download" before the line where it deletes the file, so it doesn't even do what it says it does

ok this is actually a bug and we should take more care of the dl dir... (but keeping the affected file would help us bisect the problem)

Anyway I assume NACK because having different hash between a local file and a remote file is a hard error and something really bad is happening... aka REAL SECURITY CONCERN.

If someone notice this kind of thing the first question to do would be "Is my system compromised? Is the remote system compromised? Did the makefile just have a different hash?"

Hash values are not string values just to check the integrity of the file but also to check the integrity of the entire system both client and remote.

So yes I agree... I would keep the `die` since having mismatched hash is REALLY BAD. (and just ignoring it as we have other mirror is not a good idea as (thinking absurdly) we would keep a compromised mirror and never notice that)

Reply:
I agree it would be considered a security concern, and should error out the script, the only difference is when

it would print hash mismatch, delete the file, and then print out of mirrors and die there

if there is a compromised file on a mirror, or a compromised mirror, and its not the first mirror in the mirrors list, the script would never tell us anyway, because the first attempt would be OK.

if the client is compromised, every download attempt fails, even with "warn" here, because another part of the script handles hashing the download apart from the local file if it is already there. (funny enough, I broke mkhash during testing and had that exact situation)

A larger question is whether we want to keep the local file or not if hash does not match, and if so, only the message in the die should change. However, if we REALLY consider it a security concern, I would say it should get deleted and a new download attempt happens immediately, which is what "warn" would do (but again, I understand that does assume the local file got ruined randomly and the remote file and client are not compromised, which is not likely to occur with normal usage)

the use case I actually care about is for the script to not freak out and error like there is a security concern if the file is empty, and I realize now changing from "die" to "warn" is not the only way to handle that case

I leave that decision of which change is better up to you

Reply:
Thing is what are the possibility of having an empty file?

If we really want we can add extra check on mismatched hash (like check for the remote file size and check if ours match... and then hard exit...)

To me your concern is more on checking if a file is actually corrupt and that makes sense but how common is that?

I would honestly just fix the bug with not deleting the file and be done with it :D lets not waste too much time on small thing, you already have monster pr to split ahhahah